ProcessLib/UncoupledProcessesTimeLoop.cpp

```diff
@@ -492,41 +603,109 @@ bool UncoupledProcessesTimeLoop::loop()
         INFO("=== timestep #%u (t=%gs, dt=%gs) ==============================",
              timestep, t, delta_t);
 
-        // TODO use process name
-        unsigned pcs_idx = 0;
-        for (auto& spd : _per_process_data)
+        bool coupling_iteration_converged = true;
+        for (unsigned i = 0; i < _global_coupling_max_iterations; i++)
         {
-            auto& pcs = spd->process;
-            BaseLib::RunTime time_timestep_process;
-            time_timestep_process.start();
-
-            auto& x = *_process_solutions[pcs_idx];
-
-            nonlinear_solver_succeeded = solveOneTimeStepOneProcess(
-                x, timestep, t, delta_t, *spd, *_output);
-
-            pcs.computeSecondaryVariable(t, x);
+            // TODO use process name
+            unsigned pcs_idx = 0;
+            for (auto& spd : _per_process_data)
+            {
+                auto& pcs = spd->process;
+                BaseLib::RunTime time_timestep_process;
+                time_timestep_process.start();
+
+                auto& x = *_process_solutions[pcs_idx];
+                if (i==0)
+                    pcs.preTimestep(x, t, delta_t);
+
+                if (is_staggered_coupling)
```

Comment:
This piece of code has already been split into two functions for coupling and non-coupling.